### PR TITLE
architecting bringing code from demo and new features

### PR DESCRIPTION
**Pull request by 13zebras**
## Summary
Massive PR with code moved from demo, architecting of the site, NewRoomModal component, and more.

## Learnings
- It is better and easier to use <dialog> for modals. A much better way of doing them.
- Much of the socket.io code I had tested isn't need. Just the essentials.
- Using a CSS library like DaisyUI makes styling much faster and the code more readable than just using Tailwind alone.
- Architecting a website is hard.
	
## Struggles
- useEffects. I'm a lot better with them now, but I don't 100% feel comfortable. Much better today, though, since I've had to read some of the React docs to make some improvements.

## References
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog
https://nextjs.org/docs/app/api-reference/functions/use-router
https://socket.io/docs/v4/listening-to-events/

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	const io = new Server(httpServer, {
+		cors: {
+			origin: 'http://localhost:3000',
+		},
+	})
+
+	io.on('connection', (socket) => {
+		console.log('>>> server socket.id', socket.id)
+
+		socket.on('join-room', (roomId, userName) => {
```
Comment (sandrina-p) [inline server.js]: todo: Similar to what I said yesterday, let's move all the code related to the game logic (sockets) to its own folder and here in the server we simply call the methods :)

Comment (sandrina-p) [reply server.js]: For example, I can tell you have a console.log for every socket listener. By creating a custom function (a middleware), that function can do those common to every case without duplicating code, for example:

```js
function socketOn(name, cb) {
  socket.on(name, (...args) => {
    console.log('>>>', name', args)
    cb(...args);
  });
}
```

And then you call `socketOn` wherever needed, which will automatically do the `console.log` too:

```js
socketOn('client-message', (message, roomId) => {
  // do your stuff here...
})
```

File: src/app/[room]/page.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import CardsContainer from '@/components/CardsContainer'
+
+export type ChatEvent = {
+	message: string
+	timeStamp: string
+	room: string
+}
+
+export default function UserRooms({ params }: { params: { room: string } }) {
+	const [roomId, setRoomId] = useState('')
```
Comment (sandrina-p) [inline src/app/[room]/page.tsx]: todo: you don't need a roomId  as state, it already comes from the params :) You have other components with the same mistake, pls review them all :)

File: src/app/[room]/page.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+		socket.emit('join-room', room, user)
+
+		setRoomId(room)
+		setUserName(user)
+	}, [])
+
+	return (
+		<main className='px-16 py-12 flex flex-col items-center gap-12 w-full'>
+			<h1 className='text-3xl text-green-400'>USER Scrum Diving Room</h1>
```
Comment (sandrina-p) [inline src/app/[room]/page.tsx:22]: todo: Prefer using CSS for all-caps words, it's [better for accessibility](https://www.timdunklee.com/notes/use-css-text-transform-for-uppercase-letters-a11y/)

File: src/app/[room]/page.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+			<div className='text-2xl text-zinc-500 h-[60vh] w-full flex flex-col justify-start items-center gap-8'>
+				<div className='text-2xl text-zinc-500 h-full w-full flex flex-row gap-8 justify-start items-center'>
+					<StoryPoints />
+					<CardsContainer />
+				</div>
+			</div>
+
+			{/*** Socket.io Test Section ***/}
```
Comment (sandrina-p) [inline src/app/[room]/page.tsx]: thought: I love [this type](https://preview.redd.it/iuy9fxt300811.png?auto=webp&s=156f5ffebbea5b30189eb36012e5bc7e3369c381) of comments (sarcastic :p ). Code comments should explain why something exists, not what it does.

File: src/app/host/[room]/page.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	message: string
+	timeStamp: string
+	room: string
+}
+
+export default function HostRooms({ params }: { params: { room: string } }) {
+	const [roomId, setRoomId] = useState('')
+	const [userName, setUserName] = useState('')
+
+	// biome-ignore lint/correctness/useExhaustiveDependencies: <explanation>
```
Comment (sandrina-p) [inline src/app/host/[room]/page.tsx]: todo: No, no, never ever ignore `useExhaustiveDependencies` Why did you use it?

File: src/components/NewRoomModal.tsx [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+	const [domain, setDomain] = useState('')
+	const roomUrlRef = useRef<HTMLSpanElement>(null)
+	const dialogRef = useRef<HTMLDialogElement>(null)
+	const hostRef = useRef<HTMLInputElement>(null)
+	const [hostFirstName, setHostFirstName] = useState('')
+
+	const router = useRouter()
+
+	useEffect(() => {
+		if (typeof window !== 'undefined') {
```
Comment (sandrina-p) [inline src/components/NewRoomModal.tsx]: question/todo: Why did you check if window is undefined here? In useEffect, it always exists because this hook only runs on the client/browser.

File: src/components/NewRoomModal.tsx [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+			const port = window.location.port ? `:${window.location.port}` : ''
+			setDomain(`${http}://${hostname}${port}`)
+		}
+		const hostDataLocalStorage = localStorage.getItem('scrumDivingHostData')
+		if (hostDataLocalStorage) {
+			const hostData = JSON.parse(hostDataLocalStorage)
+			if (hostData?.hostName) {
+				setHostFirstName(hostData?.hostName)
+				setHideCreateBtn(false)
+				setShowNameInput(false)
```
Comment (sandrina-p) [inline src/components/NewRoomModal.tsx]: suggestion/todo: I believe you have way too many states in this component. Each change in the DOM is a new state, but you don't need too. Here's a tip to better manage your states state: Instead of calling states based on its outcome in the DOM (hideGoBtn, hideCreateBtn, etc), try to name it based on the data that we really need to mutate the DOM and then use that data to change the DOM.

For example, in this case the state `hostFirstName` is enough to drive the DOM mutation `hideCreateBtn`, so you don't need that extra state. Instead of `disabled={hideCreateBtn}`, you can say `disabled={!!hostFirstName}`.

Does this make sense to you?

File: src/components/NewRoomModal.tsx [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+						type='button'
+						className='btn btn-warning btn-sm w-72 text-base'
+						onClick={handleCreateRoom}
+						disabled={hideCreateBtn}
+					>
+						Click to Create New Room
+					</button>
+				</div>
+			)}
+			<dialog ref={dialogRef} className='modal'>
```
Comment (sandrina-p) [inline src/components/NewRoomModal.tsx]: praise: Good usage of `dialog`! ⚡

File: src/components/NewRoomModal.tsx [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+						>
+							Copy Room URL to Clipboard
+						</button>
+					</div>
+
+					<div className='modal-action w-full gap-3 items-center'>
+						<form method='dialog'>
+							{/* if there is a button in form, it will close the modal */}
+							<button
+								disabled={hideGoBtn}
```
Comment (sandrina-p) [inline src/components/NewRoomModal.tsx]: todo: This could be a `<Link/>`, no need to be a button, or am I missing something here?

File: src/components/NewRoomModal.tsx [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+		handleShowModal()
+		console.log('%c>>> hostFirstName:', 'color: red', hostFirstName)
+		console.log('%c>>> room:', 'color: #5f0', domain, newRoomNumber)
+	}
+
+	function goToNewRoom() {
+		console.log('%c>>> user room', 'color: red', roomUrl)
+		console.log('%c>>> host room', 'color: yellow', hostRoomUrl)
+		setHideCreateBtn(true)
+		const valuesLocalStorate = JSON.stringify({
```
Comment (sandrina-p) [inline src/components/NewRoomModal.tsx]: question: Why do you decide to store in localstorage at this moment Don't you think it's better to do that earlier when  the moment the room is created when clicking to create a new room? (before opening the modal)

Comment (sandrina-p) [reply src/components/NewRoomModal.tsx]: Or even better, the moment the room is created, redirect to that page right away, and then show the link on that same page to share with colleagues at any time. Up to you to decide :)

File: src/components/socketIoTest/MessageForm.tsx [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+export default function MessageForm({ room: roomId }: { room: string }) {
+	const [message, setMessage] = useState('')
+
+	const [isLoading, setIsLoading] = useState(false)
+	const messageRef = useRef<HTMLInputElement>(null)
+
+	function handleSend(event: React.FormEvent) {
+		event.preventDefault()
+		setIsLoading(true)
```
Comment (sandrina-p) [inline src/components/socketIoTest/MessageForm.tsx]: bug: I believe you forgot to reset the `isLoading` state here ☺️

**Review by sandrina-p — CHANGES_REQUESTED** [review]

File: src/app/[room]/page.tsx (hunk 4 above)
Comment (13zebras) [reply src/app/[room]/page.tsx]: I like delineating sections of my code so I can better visually scan the document. My eye catches the comment with *** and I know where I am in a bit of code. In fact, I often make them bigger. This originally wasn't a component, it was a few dozen of lines of JSX.

File: src/app/host/[room]/page.tsx (hunk 5 above)
Comment (13zebras) [reply src/app/host/[room]/page.tsx]: Because this should only run once, and it was being updated and switching the socket connection. However, I have refactored this to include the dependencies and gotten the socket connection to not update. New PR will fix that.

File: src/app/[room]/page.tsx (hunk 2 above)
Comment (13zebras) [reply src/app/[room]/page.tsx]: Will do. These will be changed in the next PR. I've already made changes.

File: src/app/[room]/page.tsx (hunk 3 above)
Comment (13zebras) [reply src/app/[room]/page.tsx:22]: That's not the actual page title. It's just there for now so I know what page I'm on. I'm mostly focused on functionality and layout. This will 100% be gone. But I will keep that in mind for all caps!!

File: src/components/NewRoomModal.tsx (hunk 6 above)
Comment (13zebras) [reply src/components/NewRoomModal.tsx]: I will eliminate that. I only did that because that's the example I saw and then put it in the useEffect.

File: src/components/NewRoomModal.tsx (hunk 7 above)
Comment (13zebras) [reply src/components/NewRoomModal.tsx]: I've never done that, converted a string to a boolean with `!!`. But I get it. I'll work on that.

File: src/components/NewRoomModal.tsx (hunk 8 above)
Comment (13zebras) [reply src/components/NewRoomModal.tsx]: I've never used <dialog> before, but WOW, is it a billion times easier to make modals with it!! Thanks!

File: src/components/NewRoomModal.tsx (hunk 9 above)
Comment (13zebras) [reply src/components/NewRoomModal.tsx]: Visually, it's a button. So shouldn't it be a <button> tag?

File: src/components/NewRoomModal.tsx (hunk 10 above)
Comment (13zebras) [reply src/components/NewRoomModal.tsx]: The first part, yes, I could save to localStorage in the `handleCreateRoom` function. That does make more sense.

File: src/components/socketIoTest/MessageForm.tsx (hunk 11 above)
Comment (13zebras) [reply src/components/socketIoTest/MessageForm.tsx]: Oh, at the end of the function? Yep, will do!

**Comment by 13zebras** [comment]
I have all of these marked in my code as TODO's or already updated.

File: src/app/[room]/page.tsx (hunk 4 above)
Comment (sandrina-p) [reply src/app/[room]/page.tsx]: Ah I see, so this is not a normal code comment, it's a code comment to separate the areas of code. All good then ^.^

File: src/components/NewRoomModal.tsx (hunk 6 above)
Comment (sandrina-p) [reply src/components/NewRoomModal.tsx]: A daily reminder for you to always question the code you copy-paste, to help you fully understand it ^.^

File: src/components/NewRoomModal.tsx (hunk 9 above)
Comment (sandrina-p) [reply src/components/NewRoomModal.tsx]: What do you mean by `tag`? A tag is `p`, `b`, `h1`, `button`, `a`, etc...

One thing is visual, another thing is semantics. And semantically a `<button>` is different from an `<a>`.  Check this [Links and Buttons Guide - CSS Tricks](https://css-tricks.com/a-complete-guide-to-links-and-buttons/) and [Links vs ButtonsLinks - Eric](https://yatil.net/blog/buttons-vs-links)

File: src/components/socketIoTest/MessageForm.tsx (hunk 11 above)
Comment (sandrina-p) [reply src/components/socketIoTest/MessageForm.tsx]: Yes! Otherwise we can't submit a second message.

File: src/components/NewRoomModal.tsx (hunk 9 above)
Comment (13zebras) [reply src/components/NewRoomModal.tsx]: I meant to say "So it should be an `<a>` tag? Or in Next.js, a `<Link>`?"

Okay, those points in the articles make sense. I've actually reworked some things so there isn't a need for the onClick. With the onClick and all the JS in the handler, it needed to be a button.

Reworked to use `<Link>` tag. It's in new commit.

File: src/components/socketIoTest/MessageForm.tsx (hunk 11 above)
Comment (13zebras) [reply src/components/socketIoTest/MessageForm.tsx]: Done

File: src/components/NewRoomModal.tsx (hunk 7 above)
Comment (13zebras) [reply src/components/NewRoomModal.tsx]: This has been entirely reworked. I'll push in the next commit.

File: src/components/NewRoomModal.tsx (hunk 10 above)
Comment (13zebras) [reply src/components/NewRoomModal.tsx]: Done

File: server.js (hunk 1 above)
Comment (13zebras) [reply server.js]: This will be worked on after we talk. I want to confirm what you are asking.

File: src/app/[room]/page.tsx (hunk 2 above)
Comment (13zebras) [reply src/app/[room]/page.tsx]: DONE

File: src/app/[room]/page.tsx (hunk 4 above)
Comment (13zebras) [reply src/app/[room]/page.tsx]: DONE

File: src/app/host/[room]/page.tsx (hunk 5 above)
Comment (13zebras) [reply src/app/host/[room]/page.tsx]: New commits have resolved this.

File: src/components/NewRoomModal.tsx (hunk 7 above)
Comment (13zebras) [reply src/components/NewRoomModal.tsx]: DONE

File: src/components/NewRoomModal.tsx (hunk 8 above)
Comment (13zebras) [reply src/components/NewRoomModal.tsx]: Modal is gone. Reworked /host page.

File: src/components/NewRoomModal.tsx (hunk 10 above)
Comment (13zebras) [reply src/components/NewRoomModal.tsx]: This component has been deleted.

File: src/components/socketIoTest/MessageForm.tsx (hunk 11 above)
Comment (13zebras) [reply src/components/socketIoTest/MessageForm.tsx]: New commit submitted.

File: src/components/NewRoomModal.tsx (hunk 9 above)
Comment (sandrina-p) [reply src/components/NewRoomModal.tsx]: > "So it should be an <a> tag? Or in Next.js, a <Link>?"

In Next, you use `Link` when pointing to another app route, and `a` for everything else.

File: src/utils/hooks/useLocalStorage.ts [hunk 12]
```diff
@@ -0,0 +1,30 @@
+export default function useLocalStorage() {
+	// biome-ignore lint/suspicious/noExplicitAny: can pass anything
```
Comment (sandrina-p) [inline src/utils/hooks/useLocalStorage.ts]: todo: You can avoid this TS error by usingi `unknown`

File: src/utils/testData.ts [hunk 13]
```diff
@@ -0,0 +1,206 @@
+export function getRandomName() {
+	return namesArray[Math.floor(Math.random() * namesArray.length)]
+}
+
+export const namesArray = [
```
Comment (sandrina-p) [inline src/utils/testData.ts]: q/todo: Why do you export this? It's not used in any other file.

nit: The file could be `utils.ts`, start simple, no need to give specific names. And also, this ain't a file related with tests :)

File: src/services/socket.tsx [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+			socket.off(joinListener, onListenerResponse)
+			socket.off(dataListener, onListenerResponse)
+		}
+	}, [joinListener, dataListener])
+
+	return events
+}
+
+// Generic Socket.io Emitter.
+//Not sure it is needed, but keeping for now
```
Comment (sandrina-p) [inline src/services/socket.tsx:86]: Yes, you can use it above at `socketRoomEmitter` :D

File: src/services/socket.tsx [hunk 15]
```diff
@@ -0,0 +1,169 @@
+'use client'
+
+import { io } from 'socket.io-client'
+import { useEffect, useState } from 'react'
+
+export type ListenerRes = {
```
Comment (sandrina-p) [inline src/services/socket.tsx]: q: Why do you export this? I can tell you have this bad habit of exporting all stuff for no reason. Be mindful of that. Only export what's really necessary :) It's easier for in big codebases for maintainence, for example: Refactors are easier without worrying about other files. Better auto-completed with only the necessary things.

File: src/app/host/page.tsx [hunk 16]
```diff
@@ -0,0 +1,18 @@
+// import { useEffect, useState, useRef } from 'react'
+import NewRoomModal from '@/components/NewRoomModal'
```
Comment (sandrina-p) [inline src/app/host/page.tsx]: bug: You forgot to push these missing files. The app does not work on my side.

File: server.js (hunk 1 above)
Comment (sandrina-p) [reply server.js]: It’s great!

File: src/utils/hooks/useLocalStorage.ts (hunk 12 above)
Comment (13zebras) [reply src/utils/hooks/useLocalStorage.ts]: I didn't know about unknown. I need to look at that. 

I'm probably going to delete this file. I don't need a hook as Local Storage is accessed only one place, one time.

File: src/utils/testData.ts (hunk 13 above)
Comment (13zebras) [reply src/utils/testData.ts]: It is used in the CardsContainer, which I have not committed the changes yet. I guess it should be called sampleData.ts as it is a sample of names to use on the user cards.

File: src/services/socket.tsx (hunk 15 above)
Comment (13zebras) [reply src/services/socket.tsx]: I'll change that. I exported it because I had assumed I would need to use it elsewhere. I didn't realize that it would end up only being used in this file. 

I did a search in the repo, that is the only use of export that is unneeded. Everything else that is exported is used in another file.

Comment (13zebras) [reply src/services/socket.tsx]: DONE

File: src/app/host/page.tsx (hunk 16 above)
Comment (13zebras) [reply src/app/host/page.tsx]: I'll push more files right now. DONE!

File: src/components/NewRoomModal.tsx (hunk 9 above)
Comment (13zebras) [reply src/components/NewRoomModal.tsx]: I believe this one is done and can be resolved.

File: src/utils/testData.ts (hunk 13 above)
Comment (13zebras) [reply src/utils/testData.ts]: I changed the name. And the last commit has all the files, including the one that uses the namesArray. DONE.

File: src/app/host/page.tsx (hunk 16 above)
Comment (13zebras) [reply src/app/host/page.tsx]: The last commit has everything I worked on tonight and a few from last night. Plus 11 sample image files for the cards.

File: src/utils/hooks/useLocalStorage.ts (hunk 12 above)
Comment (sandrina-p) [reply src/utils/hooks/useLocalStorage.ts]: Fix it for now, and then decide later if you'll delete it or not :)

File: src/services/socket.tsx (hunk 15 above)
Comment (sandrina-p) [reply src/services/socket.tsx]: "I had assumed I would need to use it elsewhere." here's an important lesson: the tech world is full of assumptions/predictions. most of them wrong. Code what you know today and try to avoid _premature optimizations_, such as exporting functions that you don't need right now.

Feel free to resolve this conversation when you read this :)

File: src/components/CardsContainer.tsx [hunk 17]
```diff
@@ -1,9 +1,31 @@
 'use client'
 
+import UserCard from './UserCard'
+import { namesArray } from '@/utils/sampleData'
+import { sealifeImages } from '@/utils/imageLists'
+
 export default function CardsContainer() {
+	const testNames = namesArray.slice(10, 30)
```
Comment (sandrina-p) [inline src/components/CardsContainer.tsx]: nit: you forgot to rename this `testNames`

File: src/components/RoomMainUi.tsx [hunk 18]
```diff
@@ -0,0 +1,16 @@
+// HOST ROOM
+'use client'
+
+import { socketRoomEmitter } from '@/services/socket'
+
+import StoryPoints from '@/components/StoryPoints'
+import CardsContainer from '@/components/CardsContainer'
+
+export default function RoomMainUi() {
```
Comment (sandrina-p) [inline src/components/RoomMainUi.tsx:9]: suggestion: Calling this `HostRoom` :)

File: src/components/StoryPointButton.tsx [hunk 19]
```diff
@@ -0,0 +1,13 @@
+export default function StoryPointButton({
+	storyPoint,
+}: { storyPoint: number | string }) {
+	return (
+		<input
+			type='radio'
+			aria-label='radio'
```
Comment (sandrina-p) [inline src/components/StoryPointButton.tsx]: bug: Why did you add this `aria-label` with value "radio"? This is wrong, the aria-label should hold a value similar to `<label>`. Ask yourself: If this had a visual label, what would be the label? Probably `{storyPoint} points`, right?

File: src/utils/hooks/useLocalStorage.ts (hunk 12 above)
Comment (13zebras) [reply src/utils/hooks/useLocalStorage.ts]: DONE!!

File: src/components/RoomMainUi.tsx (hunk 18 above)
Comment (13zebras) [reply src/components/RoomMainUi.tsx:9]: It's not actually the Host Room. It's the Room UI for host AND user. It is the user cards and story point buttons and is common to both host and user. 

The host then gets buttons above this component, and the settings button.

File: src/components/StoryPointButton.tsx (hunk 19 above)
Comment (13zebras) [reply src/components/StoryPointButton.tsx]: I learned how to make the radio buttons into buttons, and and the example in DaisyUI had the aria-label='radio'. I don't know much about aria-label other than it used for accessibility, but that's about it. 

I'll change that to your suggestion. That makes sense!

Comment (13zebras) [reply src/components/StoryPointButton.tsx]: DONE

File: src/components/CardsContainer.tsx (hunk 17 above)
Comment (13zebras) [reply src/components/CardsContainer.tsx]: DONE
